### PR TITLE
add delegate map into staking protocol struct

### PR DESCRIPTION
**Pull request by dustinxie**
and some API to manage Delegate, to unblock Zhefeng's handler work

### CODE REVIEW

File: action/protocol/staking/protocol.go [hunk 1]
```diff
@@ -114,3 +119,36 @@ func (p *Protocol) handleRestake(ctx context.Context, act action.Action, sm prot
...
+
+func (p *Protocol) addDelegate(d *Delegate) error {
+	if err := d.Validate(); err != nil {
+		return err
+	}
+
+	if !p.hasDelegate(d.Owner) {
+		p.delegates[d.Owner.String()] = d
+	}
+	return nil
```
Comment (CoderZhi) [inline action/protocol/staking/protocol.go]: is it possible this function be called multiple times with the same key? if so, should we return error since we don't overwrite? if not, why we test "hasDelegate"

File: action/protocol/staking/protocol_test.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+		require.NoError(stk.addVote(v.d.Owner, v.add))
+		require.Equal(v.final, stk.delegates[v.d.Owner.String()].Votes)
+	}
+
+	// test delegate that does not exist
+	noName, err := address.FromBytes(make([]byte, 20))
+	require.NoError(err)
+	require.False(stk.hasDelegate(noName))
+	require.Equal(ErrInvalidAmount, errors.Cause(stk.addVote(tests[0].d.Owner, big.NewInt(-1))))
+	exceed := big.NewInt(0).Add(tests[0].final, big.NewInt(1))
```
Comment (CoderZhi) [inline action/protocol/staking/protocol_test.go]: it is hard to follow tests[0] which has been modified in line 85

File: action/protocol/staking/protocol.go [hunk 3]
```diff
@@ -114,3 +119,36 @@ func (p *Protocol) handleRestake(ctx context.Context, act action.Action, sm prot
...
+		return p.delegates[name.String()].AddVote(amount)
+	}
+	return nil
+}
+
+func (p *Protocol) subVote(name address.Address, amount *big.Int) error {
+	if p.hasDelegate(name) {
+		return p.delegates[name.String()].SubVote(amount)
+	}
+	return nil
```
Comment (CoderZhi) [inline action/protocol/staking/protocol.go]: return some error?

File: action/protocol/staking/protocol.go [hunk 4]
```diff
@@ -114,3 +119,36 @@ func (p *Protocol) handleRestake(ctx context.Context, act action.Action, sm prot
...
+
+func (p *Protocol) delDelegate(name address.Address) {
+	delete(p.delegates, name.String())
+}
+
+func (p *Protocol) addVote(name address.Address, amount *big.Int) error {
+	if p.hasDelegate(name) {
+		return p.delegates[name.String()].AddVote(amount)
+	}
+	return nil
```
Comment (CoderZhi) [inline action/protocol/staking/protocol.go]: return some error?

File: action/protocol/staking/protocol.go [hunk 5]
```diff
@@ -114,3 +119,36 @@ func (p *Protocol) handleRestake(ctx context.Context, act action.Action, sm prot
 	// TODO
 	return nil, nil
 }
+
+func (p *Protocol) hasDelegate(name address.Address) bool {
+	return p.delegates.Contains(name)
+}
+
+func (p *Protocol) addDelegate(d *Delegate) error {
```
Comment (CoderZhi) [inline action/protocol/staking/protocol.go]: could this be a function of DelegateMap? same question for the rest

File: action/protocol/staking/protocol.go [hunk 6]
```diff
@@ -114,3 +119,36 @@ func (p *Protocol) handleRestake(ctx context.Context, act action.Action, sm prot
 	// TODO
 	return nil, nil
 }
+
+func (p *Protocol) hasDelegate(name address.Address) bool {
```
Comment (CoderZhi) [inline action/protocol/staking/protocol.go]: if this is a private function, why not call p.delegates.Contains(name) directly?

File: action/protocol/staking/protocol.go (hunk 1 above)
Comment (dustinxie) [reply action/protocol/staking/protocol.go]: good point, thx

need to think the correct logic, since we discussed ysterday that we'll use 12-byte name. In this case, if an active delegate change the name, then we need to update this delegate, even though it's already in the map

File: action/protocol/staking/protocol_test.go (hunk 2 above)
Comment (dustinxie) [reply action/protocol/staking/protocol_test.go]: ok, line 85 is to get rid of the first 2 fail case, so the rest of test only use the remaining 3 valid case

File: action/protocol/staking/protocol.go (hunk 3 above)
Comment (dustinxie) [reply action/protocol/staking/protocol.go]: yes, need to think the correct logic and re-write this entire thing

File: action/protocol/staking/delegate_map.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+// Get returns the delegate by name
+func (m DelegateMap) Get(name CandName) *Delegate {
+	return m[name]
+}
+
+// Put writes the delegate into map, will overwrite if name already exists
+func (m DelegateMap) Put(d *Delegate) error {
+	if err := d.Validate(); err != nil {
```
Comment (CoderZhi) [inline action/protocol/staking/delegate_map.go]: Validate is more like a function of DelegateMap

File: action/protocol/staking/delegate_map.go [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+	"github.com/golang/protobuf/proto"
+	"github.com/pkg/errors"
+
+	"github.com/iotexproject/iotex-core/action/protocol/staking/stakingpb"
+)
+
+type (
+	// DelegateMap is a map of delegates using owner's 12-byte name as key
+	DelegateMap map[CandName]*Delegate
```
Comment (yutongp) [inline action/protocol/staking/delegate_map.go]: I think you would need two maps. one is CandName to OwnerAddress mapping and one is OwnerAddress to Delegate instance mapping
